File: public/java/051-climbing-stairs.java
/*
 * LeetCode Problem #70: Climbing Stairs
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/climbing-stairs/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Recursive Memoization (Top-Down DP) | O(n) time | O(n) space
    // EXPLAIN: Cache overlapping subproblems of f(n)=f(n-1)+f(n-2) to avoid re-computation.
    // WHEN: Use when top-down thinking is natural and recursion depth is acceptable.

    public int climbStairs_memo(int n) {
        int[] memo = new int[n + 1];
        return helper(n, memo);
    }

    private int helper(int n, int[] memo) {
        if (n <= 1) return 1;
        if (memo[n] != 0) return memo[n];
        memo[n] = helper(n - 1, memo) + helper(n - 2, memo);
        return memo[n];
    }

    // APPROACH 2: Bottom-Up DP | O(n) time | O(n) space
    // EXPLAIN: Fill dp array from base cases up; dp[i] = dp[i-1] + dp[i-2].
    // WHEN: Use as the straightforward iterative DP solution — easy to explain.

    public int climbStairs_dp(int n) {
        if (n <= 1) return 1;
        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }

    // APPROACH 3: Space-Optimized DP | O(n) time | O(1) space
    // EXPLAIN: Keep only the last two Fibonacci values; discard the full array.
    // WHEN: Use in interviews as the optimal solution — constant space.

    public int climbStairs(int n) {
        if (n <= 1) return 1;
        int prev2 = 1, prev1 = 1;
        for (int i = 2; i <= n; i++) {
            int cur = prev1 + prev2;
            prev2 = prev1;
            prev1 = cur;
        }
        return prev1;
    }

    // APPROACH 4: Matrix Exponentiation | O(log n) time | O(1) space
    // EXPLAIN: Represent Fibonacci as matrix power [[1,1],[1,0]]^n; use fast exponentiation.
    // WHEN: Use for very large n where even O(n) is too slow — academic/competitive programming.

    public int climbStairs_matrix(int n) {
        if (n <= 2) return n;
        long[][] base   = {{1, 1}, {1, 0}};
        long[][] result = matPow(base, n);
        return (int) result[0][0];
    }

    private long[][] matPow(long[][] m, int n) {
        if (n == 1) return m;
        if (n % 2 == 0) {
            long[][] half = matPow(m, n / 2);
            return matMul(half, half);
        }
        return matMul(m, matPow(m, n - 1));
    }

    private long[][] matMul(long[][] A, long[][] B) {
        long[][] C = new long[2][2];
        for (int i = 0; i < 2; i++)
            for (int j = 0; j < 2; j++)
                for (int k = 0; k < 2; k++)
                    C[i][j] += A[i][k] * B[k][j];
        return C;
    }

    // APPROACH 5: Mathematical Formula (Binet's Formula) | O(1) time | O(1) space
    // EXPLAIN: Closed-form Fibonacci using phi = (1+√5)/2; answer is F(n+1) rounded.
    // WHEN: Mathematical curiosity — floating-point precision limits practical use for large n.

    public int climbStairs_formula(int n) {
        double sqrt5 = Math.sqrt(5);
        double phi   = (1 + sqrt5) / 2;
        double psi   = (1 - sqrt5) / 2;
        return (int) Math.round((Math.pow(phi, n + 1) - Math.pow(psi, n + 1)) / sqrt5);
    }
}

// Made with Bob

File: public/python/051-climbing-stairs.py
# LeetCode Problem #70: Climbing Stairs
# Difficulty: Easy
# Link: https://leetcode.com/problems/climbing-stairs/

from typing import List
import math

# ─────────────────────────────────────────────
# APPROACH 1: Recursive Memoization (Top-Down DP) | O(n) time | O(n) space
# EXPLAIN: Cache f(n)=f(n-1)+f(n-2) top-down using lru_cache to avoid exponential blowup.
# WHEN: First instinct in interviews — easy to derive from the recurrence.

from functools import lru_cache

def climbStairs_memo(n: int) -> int:
    @lru_cache(maxsize=None)
    def dp(i):
        if i <= 1:
            return 1
        return dp(i - 1) + dp(i - 2)
    return dp(n)


# APPROACH 2: Bottom-Up DP | O(n) time | O(n) space
# EXPLAIN: Build a table from base cases up to n; dp[i] = dp[i-1] + dp[i-2].
# WHEN: When you want explicit tabulation that is easy to trace step by step.

def climbStairs_dp(n: int) -> int:
    if n <= 1:
        return 1
    dp = [0] * (n + 1)
    dp[0], dp[1] = 1, 1
    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return dp[n]


# APPROACH 3: Space-Optimised DP (Two Variables) | O(n) time | O(1) space
# EXPLAIN: Only the previous two values matter; roll them forward without an array.
# WHEN: Production-ready — minimal memory, identical speed.

def climbStairs_space_opt(n: int) -> int:
    if n <= 1:
        return 1
    prev2, prev1 = 1, 1
    for _ in range(2, n + 1):
        prev2, prev1 = prev1, prev1 + prev2
    return prev1


# APPROACH 4: Matrix Exponentiation | O(log n) time | O(1) space
# EXPLAIN: [[1,1],[1,0]]^n gives Fibonacci; fast binary exponentiation achieves O(log n).
# WHEN: For very large n where O(n) is too slow — competitive programming / academic.

def climbStairs_matrix(n: int) -> int:
    if n <= 2:
        return n

    def mat_mul(A, B):
        return [
            [A[0][0]*B[0][0] + A[0][1]*B[1][0], A[0][0]*B[0][1] + A[0][1]*B[1][1]],
            [A[1][0]*B[0][0] + A[1][1]*B[1][0], A[1][0]*B[0][1] + A[1][1]*B[1][1]],
        ]

    def mat_pow(m, p):
        if p == 1:
            return m
        if p % 2 == 0:
            half = mat_pow(m, p // 2)
            return mat_mul(half, half)
        return mat_mul(m, mat_pow(m, p - 1))

    base = [[1, 1], [1, 0]]
    result = mat_pow(base, n)
    return result[0][0]


# APPROACH 5: Mathematical Formula (Binet's) | O(1) time | O(1) space
# EXPLAIN: Closed-form Fibonacci via phi = (1+√5)/2; answer = F(n+1) rounded.
# WHEN: Mathematical elegance only — floating-point limits precision for large n.

def climbStairs_formula(n: int) -> int:
    sqrt5 = math.sqrt(5)
    phi   = (1 + sqrt5) / 2
    psi   = (1 - sqrt5) / 2
    return round((phi**(n + 1) - psi**(n + 1)) / sqrt5)


# ── quick tests ──────────────────────────────────────────────────────────────
if __name__ == '__main__':
    cases = [(1, 1), (2, 2), (3, 3), (4, 5), (5, 8), (10, 89)]
    for fn in (climbStairs_memo, climbStairs_dp, climbStairs_space_opt,
               climbStairs_matrix, climbStairs_formula):
        for n, expected in cases:
            result = fn(n)
            assert result == expected, f'{fn.__name__}({n}) = {result}, expected {expected}'
    print('All tests passed.')

# Made with Bob
